src/rgw/rgw_common.cc

```diff
@@ -1152,18 +1153,22 @@ bool verify_bucket_permission(struct req_state * const s, const uint64_t op)
                                   op);
 }
 
+// Authorize anyone permitted by the policy and the bucket owner
+// unless explicitly denied by the policy.
+
 int verify_bucket_owner_or_policy(struct req_state* const s,
 				  const uint64_t op)
 {
-  if (s->iam_policy) {
-    if (s->iam_policy->eval(s->env, *s->auth.identity, op,
-			    ARN(s->bucket)) == Effect::Allow) {
-      return 0;
-    }
-  } else if (s->auth.identity->is_owner_of(s->bucket_owner.get_id())) {
-    return 0;
-  }
-  return -EACCES;
+  auto e = maybe_do_or(s->iam_policy,
```

Comment:
I like this idea in theory, but in this particular case, the functional abstraction seems a lot less explicit, and I'm kind of suspecting, maybe adds a bit of overhead?